rust/blockstore/tests/blockfile_writer_test.rs

```diff
@@ -0,0 +1,469 @@
+                            .writer
+                            .delete::<&str, String>(prefix.as_str(), key.as_str()),
+                    )
+                    .unwrap();
+                }
+                Transition::Commit => {
+                    let id = state.writer.id();
+                    let flusher = block_on(state.writer.commit::<&str, String>()).unwrap();
+                    block_on(flusher.flush::<&str, String>()).unwrap();
+
+                    state.last_blockfile_id = Some(id);
+                    state.writer = block_on(
+                        state.provider.write::<&str, String>(
+                            BlockfileWriterOptions::new()
+                                .set_mutation_ordering(
+                                    ref_state.generated_mutation_ordering.clone(),
+                                )
+                                .fork(id),
+                        ),
+                    )
+                    .unwrap();
+                }
+            }
+
+            state
+        }
+
+        fn check_invariants(
+            state: &Self::SystemUnderTest,
+            ref_state: &<Self::Reference as ReferenceStateMachine>::State,
+        ) {
+            if ref_state.last_commit.is_none() || state.last_blockfile_id.is_none() {
+                return;
+            }
+
+            let ref_last_commit = ref_state.last_commit.as_ref().unwrap();
+            let last_blockfile_id = state.last_blockfile_id.unwrap();
+
+            let reader = block_on(state.provider.open::<&str, &str>(&last_blockfile_id)).unwrap();
+
+            // Check count
+            assert_eq!(block_on(reader.count()).unwrap(), ref_last_commit.len());
+
+            // Check that entries are ordered and match expected
+            if let Some(min_key) = ref_last_commit.keys().next() {
+                let all_entries =
+                    block_on(reader.get_gte(min_key.0.as_str(), min_key.1.as_str())).unwrap();
+
+                for (blockfile_entry, expected_entry) in
+                    all_entries.iter().zip(ref_last_commit.iter())
+                {
+                    assert_eq!(blockfile_entry.0, expected_entry.0 .1); // key matches
+                    assert_eq!(blockfile_entry.1, expected_entry.1); // value matches
+                }
+            }
+
+            // Check that all blocks are under the max block size
+            let mut checked_at_least_one_block = false;
+            for file in std::fs::read_dir(state.storage_dir.path().join("block")).unwrap() {
+                let file = file.unwrap();
+                let path = file.path();
+                if path.is_file() {
+                    let block = Block::load_with_validation(path.to_str().unwrap(), Uuid::new_v4())
+                        .unwrap();
+                    assert!(block.get_size() <= ref_state.generated_max_block_size_bytes);
+                    checked_at_least_one_block = true;
+                }
+            }
```

Comment:
maybe add a is_valid method or similar on the reader instead, also something like that for the sparse index

Reply:
doing this and gating the method with `#[cfg(test)]` doesn't work because this is [an integration test](https://www.reddit.com/r/rust/comments/ny6k3f/cfgtest_doesnt_take_affect_when_running/)

would you prefer that the method is public for all build targets or that we move this test back into `src/`?

Reply:
I think public is fine. Also fine with moving the test into the src/ the only thing having it outside is force a library level checking. So slight preference to keeping the test here.